hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/move/MoveManagerImpl.java

```diff
@@ -0,0 +1,750 @@
+      DatanodeDetails dn = cop.getTarget();
+      if (opType.equals(PendingOpType.ADD) && mdnp.getTgt().equals(dn)) {
+        if (pendingMoveFuture.containsKey(containerID)) {
+          try {
+            handleSuccessfulAdd(containerID);
+          } catch (ContainerNotFoundException | NodeNotFoundException |
+                   ContainerReplicaNotFoundException e) {
+            LOG.warn("Can not handle successful Add for move : {}", e);
+          }
+        } else {
+          //if MoveManager is reinitialize, pendingMoveOps will be restored,
+          //but pendingMoveFuture not. so there will be a case that
+          //container is in pendingMoveOps, but not in pendingMoveFuture.
+          // we just complete this move for now.
+          // also, we can continue this move, to be discussed.
+          completeMove(containerID, null);
+        }
+      } else if (opType.equals(PendingOpType.DELETE)
+          && mdnp.getSrc().equals(dn)) {
+        completeMove(containerID, MoveResult.COMPLETED);
+      }
+    }
+  }
+
+   /**
+   * notify move manager that a container op has been Expired.
+   *
+   * @param cop ContainerReplicaOp
+   * @param containerID ContainerID for which to complete
+   */
+  @Override
+  public void notifyContainerOpExpired(ContainerReplicaOp cop,
+      ContainerID containerID) {
+    if (!running) {
+      return;
+    }
+
+    MoveDataNodePair mdnp = pendingMoveOps.get(containerID);
+    if (mdnp != null) {
+      PendingOpType opType = cop.getOpType();
+      DatanodeDetails dn = cop.getTarget();
+      if (opType.equals(PendingOpType.ADD) && mdnp.getTgt().equals(dn)) {
+        completeMove(containerID, MoveResult.REPLICATION_FAIL_TIME_OUT);
+      } else if (opType.equals(PendingOpType.DELETE) &&
+          mdnp.getSrc().equals(dn)) {
+        completeMove(containerID, MoveResult.DELETION_FAIL_TIME_OUT);
+      }
+    }
+  }
+
+  private void handleSuccessfulAdd(final ContainerID cid)
+      throws ContainerNotFoundException,
+      ContainerReplicaNotFoundException, NodeNotFoundException {
+    MoveDataNodePair movePair = pendingMoveOps.get(cid);
+    if (movePair == null) {
+      return;
+    }
+    final DatanodeDetails src = movePair.getSrc();
+    final Set<ContainerReplica> replicaSet =
+        containerManager.getContainerReplicas(cid);
+    if (!replicaSet.stream()
+        .anyMatch(r -> r.getDatanodeDetails().equals(src))) {
+      // if the target is present but source disappears somehow,
+      // we can consider move is successful.
+      completeMove(cid, MoveResult.COMPLETED);
+      return;
+    }
+
+    final NodeStatus nodeStatus = nodeManager.getNodeStatus(src);
+    if (!nodeStatus.isInService()) {
+      completeMove(cid, MoveResult.DELETION_FAIL_NODE_NOT_IN_SERVICE);
+      return;
+    }
+    if (!nodeStatus.isHealthy()) {
+      completeMove(cid, MoveResult.DELETION_FAIL_NODE_UNHEALTHY);
+      return;
+    }
+
+    final ContainerInfo cif = containerManager.getContainer(cid);
```

Comment:
I feel this logic needs to go via `ReplicationManager.processOverReplicatedContainer` so we are not duplicating the OverRep handling logic here. We can change this interface a bit to make it more suitable and allow for passing some "preferred delete" replica.

We still need to add the handlers for Ratis and make the interface work with both of them.

I am also not sure these processUnderReplicatedContainer / processOverReplicatedContainer method need to be given a ContainerHealthResult object as their input. That is what is on the underReplicated queue, but I am not sure if we could get away with just passing the containerInfo instead.

Reply:
> I feel this logic needs to go via ReplicationManager.processOverReplicatedContainer

agree, but it probably brings much refactor work. we can create a separate jira for refactoring

> but I am not sure if we could get away with just passing the containerInfo instead.

underReplicated queue is a little more complex than other two queues, since it is a priority queue, which need some additional informations(WeightedRedundancy and RequeueCount) to sort the items.

Reply:
I don't think there is much refactoring to do here.

`ReplicationManager.processOverReplicatedContainer` processes the container for over replication and then returns a list of commands (ie deleteContainer commands). The Ratis side of things is not implemented yet, but Siddhant is planning to do that starting this week. The part we would need to change, is to pass a "hint" into the processOverReplication that we want to delete a certain replica if it makes sense to do so. It means all the placement checks etc are covered in one place in that handler, and not duplicate in the move manager too.